mlir/docs/Tutorials/MlirOpt.md

```diff
@@ -0,0 +1,276 @@
+if you try to run a pass that is anchored on some other op
+inside `--pass-pipeline="builtin.module(pass-name)"`,
+it will not run.
+
+Multiple passes can be chained together
+by providing the pass names in a comma-separated list
+in the `--pass-pipeline` string,
+e.g.,
+`--pass-pipeline="builtin.module(pass1,pass2)"`.
+The passes will be run sequentially.
+
+To use passes that have nontrivial anchoring,
+the appropriate level of nesting must be specified
+in the pass pipeline.
+For example, consider the following IR which has the same redundant code,
+but in two different levels of nesting.
+
+```mlir
+module {
+  module {
+    func.func @func1(%arg0: i32) -> i32 {
+      %0 = arith.addi %arg0, %arg0 : i32
+      %1 = arith.addi %arg0, %arg0 : i32
+      %2 = arith.addi %0, %1 : i32
+      func.return %2 : i32
+    }
+  }
+
+  gpu.module @gpu_module {
+    gpu.func @func2(%arg0: i32) -> i32 {
+      %0 = arith.addi %arg0, %arg0 : i32
+      %1 = arith.addi %arg0, %arg0 : i32
+      %2 = arith.addi %0, %1 : i32
+      gpu.return %2 : i32
+    }
+  }
+}
+```
+
+The following pipeline runs `cse` (common subexpression elimination)
+but only on the `func.func` inside the two `builtin.module` ops.
+
+```bash
+build/bin/mlir-opt mlir/test/Examples/mlir-opt/ctlz.mlir --pass-pipeline='
+    builtin.module(
+        builtin.module(
+            func.func(cse,canonicalize),
+            convert-to-llvm
+        )
+    )'
+```
+
+The output leaves the `gpu.module` alone
+
+```mlir
+module {
+  module {
+    llvm.func @func1(%arg0: i32) -> i32 {
+      %0 = llvm.add %arg0, %arg0 : i32
+      %1 = llvm.add %0, %0 : i32
+      llvm.return %1 : i32
+    }
+  }
+  gpu.module @gpu_module {
+    gpu.func @func2(%arg0: i32) -> i32 {
+      %0 = arith.addi %arg0, %arg0 : i32
+      %1 = arith.addi %arg0, %arg0 : i32
+      %2 = arith.addi %0, %1 : i32
+      gpu.return %2 : i32
+    }
+  }
+}
+```
+
+For a spec of the pass-pipeline textual description language,
+see [the docs](/docs/PassManagement/#textual-pass-pipeline-specification).
+For more general information on pass management, see [Pass Infrastructure](/docs/PassManagement/#).
+
+## Useful CLI flags
+
+- `--debug` prints all debug information produced by `LLVM_DEBUG` calls.
+- `--debug-only="my-tag"` prints only the debug information produced by `LLVM_DEBUG`
+  in files that have the macro `#define DEBUG_TYPE "my-tag"`.
+  This often allows you to print only debug information associated with a specific pass.
+    - `"greedy-rewriter"` only prints debug information
+      for patterns applied with the greedy rewriter engine.
+    - `"dialect-conversion"` only prints debug information
+      for the dialect conversion framework.
+ - `--emit-bytecode` emits MLIR in the bytecode format.
+ - `--mlir-pass-statistics` print statistics about the passes run.
+    These are generated via [pass statistics](/docs/PassManagement/#pass-statistics).
+ - `--mlir-print-ir-after-all` prints the IR after each pass.
```

Comment:
```suggestion
 - `--mlir-print-ir-after-all` prints the IR after each pass,
     and ` --mlir-print-ir-tree-dir` dumps it on disk.
```

The` --mlir-print-ir-tree-dir` is much more practical!

(There is also the `  --mlir-print-ir-before=<pass-arg> ` variant)

Reply:
> When printing the IR before/after a pass, print file tree rooted at this directory

I can't seem to get this flag to to anything, and the help string doesn't help me understand what it is supposed to do....

Reply:
Seems like we should improve the help!
It is to be used in conjunction with the other IR printing  options.

For example:
```
$ ./bin/mlir-opt ../mlir/test/Transforms/inlining.mlir  --pass-pipeline="builtin.module(func.func(canonicalize, cse))"  --mlir-print-ir-tree-dir=/tmp/foo --mlir-print-ir-after-all --mlir-print-ir-before-all -o /dev/null

$ find /tmp/foo | sort
/tmp/foo
/tmp/foo/builtin_module_no-symbol-name
/tmp/foo/builtin_module_no-symbol-name/0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/1_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_arg/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multiblock/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_convert_callee_fn_multi_res/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_double_square_complex/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_2_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_external/0_3_cse.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_0_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_1_canonicalize.mlir
/tmp/foo/builtin_module_no-symbol-name/func_func_func_with_arg/0_2_cse.mlir
....
```

Reply:
Done.